### ISSUE TITLE
Unnecessary burn-nn usage in the burn-store crate

### ISSUE THREAD

**Issue by softmaximalist** [tracel-ai/burn#4576]
### Issue
I am trying to implement a loss function in the `burn-nn` crate. The loss function depends on a pretrained model. I am planning to load the pretrained weights from pytorch. However, since `burn-store` depends on `burn-nn` (`[dependencies]`), using `burn-store` in the `burn-nn` crate (as `[dependencies]`) would result in circular dependency. 

The only non-test and non-benchmarking usage of `burn-nn` in the `burn-store` crate is in the [`crates/burn-store/src/adapter.rs`](https://github.com/tracel-ai/burn/blob/main/crates/burn-store/src/adapter.rs#L24) file. After looking through the code, I think I understand how `burn-nn` is used in here.

### My Understanding
Basically, `BatchNorm`, `Linear`, etc are the actual names of the neural network components/modules available in Burn (i.e., `burn-nn` crate) that developers can use to build models. Because the `burn-store` adapter relies on hardcoded string text like `"Struct:Linear"` to know how to load weights, a developer added the `use` statement as an alarm system. If a maintainer or contributor renames a neural network component/module in the `burn-nn` crate (e.g., renaming `Linear` to `NewLinear`), then the import of `Linear` in the `adapter.rs` file should result in a compile error. This would remind the maintainers to update the hardcoded strings in the `burn-store` crate. 

### My Proposed Solution
If we can somehow get rid of this specific usage of `burn-nn` in the `burn-store` crate, then we should be able to move `burn-nn` from `[dependencies]` to `[dev-dependencies]` in the `Cargo.toml` file of the `burn-store` crate. Then, adding `burn-store` to `[dependencies]` for `burn-nn` should not result in a circular dependency that the compiler rejects. 

Thus, we should add a new test case to the `crates/burn-store/src/adapter.rs` file as follows:
```rust
#[test]
fn test_module_names_match_burn_nn() {
    // If these types are renamed or moved in `burn-nn`, this test will fail to compile.
    // This use statement should be enough to replicate the current check/alarm system.
    use burn_nn::{BatchNorm, GroupNorm, LayerNorm, Linear};

    // Use `stringify!` to ensure the hardcoded strings perfectly 
    // match the actual names of the structs in `burn-nn`.
    // These assert statements work as extra checks that should remind maintainers more
    // clearly that the hardcoded strings needs get updated.
    assert_eq!(module_names::LINEAR, format!("Struct:{}", stringify!(Linear)));
    assert_eq!(module_names::BATCH_NORM, format!("Struct:{}", stringify!(BatchNorm)));
    assert_eq!(module_names::LAYER_NORM, format!("Struct:{}", stringify!(LayerNorm)));
    assert_eq!(module_names::GROUP_NORM, format!("Struct:{}", stringify!(GroupNorm)));
}
```

### Conclusion
Let me know if I am understanding the usage of `burn-nn` in the `crates/burn-store/src/adapter.rs` file correctly and whether my proposed solution is good or not. If everything is good, should I create a new PR for these changes or should I include these changes in my PR for the Gram Matrix Loss implementation?

**Comment by antimora** [comment]
Yes, it sounds good solution. It was mainly to get attention if module changes somehow.

**Comment by softmaximalist** [comment]
Thank you for the confirmation. Should I open a new PR or include the changes in my Gram Matrix Loss PR?

**Comment by antimora** [comment]
New PR can be faster to verify
